### DISCUSSION TITLE
Is there a way to allow more then 2 decimal places for prices?

### DISCUSSION THREAD

**Discussion by toastedtoast** [saleor/saleor#9576] [General] [answered]
Hi 👋,

for some of our products we want our customer to be able to individually select the weight / amount of the product. E.g. ordering 220g of rice. 

### What I want to achieve 
The basic idea is to configure the product and the price based on 1 gramm meaning that if the user wants to buy 220g of rice I am adding the quantity of 220 of this product to the checkout. Assuming a kg of rice is 4,98€ I'd like to set the price of the product to 0.00498€. Therefore the total prices calculates to 1,10€ (rounded)

### What my problem is
Saleor dashboard tells me that `Value cannot have more than 2 decimal places.` which makes perfectly sense for prices of course, but is a showstopper for my usecase.

### What I want to ask
Is there a way to allow prices with more then 2 decimal places? Or do you have other suggestions on how to approach this use case with saleor?

Kind regards,
Marco

**Comment by timuric** [comment, marked as answer]
@toastedtoast our API doesn't work with such precision, and it would not be supported by any payment gateway anyway. A possible workaround would be to use custom pricing webhooks and attributes. Alternatively, you can fork both dashboard and core API to change precision

**Reply by toastedtoast** [reply]
Hi @timuric,
definitely makes sense for most use cases to only support 2 decimals so I perfectly understand that it's not intended to be flexible.
I'd prefer to avoid forking the whole project but I do also not really understand your suggestion regarding custom pricing webhook and attributes.

The workaround I currently have in mind looks something like this:

1. Add attributes to products about the factor / amount for the price. E.g. if the product price says 0,12 EUR and the factor is 10 the price stands for 10 gramm. So the client adds 10x the product to the checkout for 100g (1,20EUR). (I know this sounds confusing).
2. In the shop frontend I always consider this factor so the user always sees 100g (instead of 10x)
3. Checkout and Payment just works as expected as it's based on amount (10x) and fullfilment (backoffice process) just need to understand the factor, too so the user gets 100g delivered.
4. For creating emails and invoices I also need to consider the factor whenever a product uses it. (Do I need to implement custom plugins for these or do you think I can fulfil this just using handlebars 🤔(Have not touched the email templates yet...))

Is something like this the workaround you mention (webhooks and attributes) or is there a more subtle, smarter approach?

Thank you for your help!

**Reply by patrys** [reply]
At one point in time, we discussed allowing staff to set both a price and a quantity denominator for products (for example, "$5.20 for 100"). The API would still allow any quantity to be ordered, and the final price would be determined by dividing the quantity ordered by the denominator.

I liked that idea, and while we didn't get around to implementing it, I think it would make it useful for items typically priced by weight (like fresh produce) and cheap items usually sold in large quantities (like screws and washers).

**Reply by toastedtoast** [reply]
That'd be exactly what we need! 
There is currently no plan from your side to add this feature anytime soon, right?

**Reply by patrys** [reply]
There are no immediate plans, but it's valuable input to know it would help in your case.

**Reply by toastedtoast** [reply]
Thanks, great to hear.
One thing that got introduced with 3.2.0 is those custom prices for checkout items [PR9393](https://github.com/saleor/saleor/pull/9393), which actually might also help in my case although I am not 100% certain if I got the feature right, yet.

So assuming I have a product which I set the price to 1.20EUR and I use an attribute which tells the frontend that this is the price for lets say 100g. The user now sets 200(g) as a quantity and instead of 240EUR (200*1.2EUR) I set the custom price to 2.4 for the checkout item.

 That way I do get the quantity of 200 while the calculation is based on a 1/100. 
 I assume this checkout price is also passed on to the Order confirmation and invoice generation, right?

Do I understand this feature right? Do you think this might be an alternative to my workaround? 
That would be significantly easier to handle that my workaround 😅

Thanks.

**Reply by toastedtoast** [reply]
@patrys @timuric 
I hope you are still with me as I really want to solve this with saleor and I hope you can add some of your thoughts.

I am basically able to do the calculation in the client based on attributes and the checkout custom price feature as mentioned above. One thing I learned is that this custom price is per item (which generally makes sense). So in my situation I want to add quantity 500 of rice with a custom price which reflects the calculate price per gramm. Of course the price per gramm is pretty low e.g. 0.0039EUR

```
{
  quantity: 500
  variantId: "XYZ"
  price: 0.0039
} 
```

Of course the thing that happens is: Price gets trimmed to 2 decimals => 0.00EUR. 
Is there any way to tell saleor to not trimm the price override for checkout items? 

I assume you'll say no but still maybe you guys have an idea how I can handle this?
Alternatively one thing that would help me a lot is to be able to assign decimal quantities eg. 0.5 so I can use prices per kilogram ;)

**Comment by djkato** [comment]
@toastedtoast Hello, did you ever get around to solving this? I'm trying to port a website that sells tiny things like buttons, broches etc and our price also has to be to at least 3 decimal precision.

**Reply by toastedtoast** [reply]
Unfortunately I cannot provide any solution for you. We also put our project on ice for now. 
My approach was nothing I'd like to use in production. 🙈
